### PR TITLE
Fix persistence allocations part 2.

### PR DESCRIPTION
**Pull request by F0b0s**
## Changes

Next optimization of persistence allocation that was started in https://github.com/akkadotnet/akka.net/pull/5505
Now excessive object allocation was eliminated in Eventsourced class when using linked list. Benchmark shows kinda 10% lower memory usage.

## Checklist

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).

### Latest `dev` Benchmarks 

BenchmarkDotNet=v0.13.2, OS=macOS Monterey 12.5.1 (21G83) [Darwin 21.6.0]
Apple M1 2.40GHz, 1 CPU, 8 logical and 8 physical cores
.NET SDK=7.0.101
  [Host] : .NET 7.0.1 (7.0.122.56804), X64 RyuJIT SSE4.2 DEBUG

Toolchain=InProcessEmitToolchain  InvocationCount=1  UnrollFactor=1  

|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |   StdDev |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|---------:|----------:|----------:|------------:|
| WriteToPersistence |                1 |           100 |    696.0 us |  44.63 us | 122.9 us |         - |         - |   146.88 KB |
| WriteToPersistence |               10 |           100 |  1,704.8 us |  44.52 us | 129.2 us |         - |         - |  1420.27 KB |
| WriteToPersistence |              100 |           100 | 23,525.9 us | 458.53 us | 657.6 us | 2000.0000 | 1000.0000 | 13786.62 KB |

### This PR's Benchmarks

BenchmarkDotNet=v0.13.2, OS=macOS Monterey 12.5.1 (21G83) [Darwin 21.6.0]
Apple M1 2.40GHz, 1 CPU, 8 logical and 8 physical cores
.NET SDK=7.0.101
  [Host] : .NET 7.0.1 (7.0.122.56804), X64 RyuJIT SSE4.2 DEBUG

Toolchain=InProcessEmitToolchain  InvocationCount=1  UnrollFactor=1  

|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |    StdDev |      Median |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|----------:|------------:|----------:|----------:|------------:|
| WriteToPersistence |                1 |           100 |    654.4 us |  31.10 us |  87.20 us |    647.0 us |         - |         - |    131.8 KB |
| WriteToPersistence |               10 |           100 |  1,786.2 us |  75.45 us | 215.27 us |  1,712.7 us |         - |         - |  1287.63 KB |
| WriteToPersistence |              100 |           100 | 23,127.9 us | 454.94 us | 637.76 us | 23,165.5 us | 2000.0000 | 1000.0000 | 12685.84 KB |

### CODE REVIEW

File: src/core/Akka.Persistence/JournalProtocol.cs [hunk 1]
```diff
@@ -227,22 +216,19 @@ public WriteMessages(IEnumerable<IPersistentEnvelope> messages, IActorRef persis
...
-
-       
+        
         public bool Equals(WriteMessages other)
         {
             if (ReferenceEquals(other, null)) return false;
             if (ReferenceEquals(this, other)) return true;
 
-            return Equals(ActorInstanceId, other.ActorInstanceId)
+            return ActorInstanceId == other.ActorInstanceId
```
Comment (to11mtm) [inline src/core/Akka.Persistence/JournalProtocol.cs:225]: good catch

**Review by to11mtm — APPROVED** [review]

File: src/core/Akka.Persistence/JournalProtocol.cs (hunk 1 above)
Comment (F0b0s) [reply src/core/Akka.Persistence/JournalProtocol.cs:225]: Rider highligths such cases

File: src/core/Akka.Persistence/Eventsourced.Recovery.cs [hunk 2]
```diff
@@ -334,11 +332,11 @@ private void FlushBatch()
         {
             if (_eventBatch.Count > 0)
             {
-                foreach (var p in _eventBatch.Reverse())
+                foreach (var p in _eventBatch)
                 {
                     _journalBatch.Add(p);
                 }
-                _eventBatch = new LinkedList<IPersistentEnvelope>();
+                _eventBatch.Clear();
```
Comment (Aaronontheweb) [inline src/core/Akka.Persistence/Eventsourced.Recovery.cs:339]: Nice catch.

File: src/core/Akka.Persistence/Eventsourced.cs [hunk 3]
```diff
@@ -303,7 +303,7 @@ public void Persist<TEvent>(TEvent @event, Action<TEvent> handler)
 
             _pendingStashingPersistInvocations++;
             _pendingInvocations.AddLast(new StashingHandlerInvocation(@event, o => handler((TEvent)o)));
-            _eventBatch.AddFirst(new AtomicWrite(new Persistent(@event, persistenceId: PersistenceId,
+            _eventBatch.AddLast(new AtomicWrite(new Persistent(@event, persistenceId: PersistenceId,
```
Comment (Aaronontheweb) [inline src/core/Akka.Persistence/Eventsourced.cs:306]: Why this change @F0b0s ?

File: src/core/Akka.Persistence/Journal/AsyncWriteJournal.cs [hunk 4]
```diff
@@ -341,7 +341,7 @@ private void HandleWriteMessages(WriteMessages message)
              */
             var self = Self;
             _resequencerCounter += message.Messages.Aggregate(1, (acc, m) => acc + m.Size);
-            var atomicWriteCount = message.Messages.OfType<AtomicWrite>().Count();
+            var atomicWriteCount = message.Messages.Count(x => x is AtomicWrite);
```
Comment (Aaronontheweb) [inline src/core/Akka.Persistence/Journal/AsyncWriteJournal.cs:344]: LGTM

**Review by Aaronontheweb — COMMENTED** [review]
Have a question about changing the order of the stashing invocations on the persistent actors, but otherwise it looks good.

File: src/core/Akka.Persistence/Eventsourced.cs (hunk 3 above)
Comment (F0b0s) [reply src/core/Akka.Persistence/Eventsourced.cs:306]: The only place where _eventBatch is iterated is `FlushBatch` method and it use `Reverse` like this:
```
                foreach (var p in _eventBatch.Reverse())
                {
                    _journalBatch.Add(p);
                }
```
it end's up in excessive memory allocation. I changed `AddFirst` to `AddLast` and there is no need for reverse anymore.
Actually i don't know why there were used `AddFirst` and `Reverse`.

Comment (Aaronontheweb) [reply src/core/Akka.Persistence/Eventsourced.cs:306]: Good catch - makes sense.

**Review by Aaronontheweb — APPROVED** [review]
LGTM
